subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultPluginManager.java

```diff
@@ -116,37 +120,17 @@ public void apply(Class<?> type) {
         doApply(pluginRegistry.inspect(type));
     }
 
-    private void doApply(PluginImplementation<?> plugin) {
-        PluginId pluginId = plugin.getPluginId();
-        String pluginIdStr = pluginId == null ? null : pluginId.toString();
+    private void doApply(final PluginImplementation<?> plugin) {
         Class<?> pluginClass = plugin.asClass();
         ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
         try {
             Thread.currentThread().setContextClassLoader(pluginClass.getClassLoader());
             if (plugin.getType().equals(PotentialPlugin.Type.UNKNOWN)) {
                 throw new InvalidPluginException("'" + pluginClass.getName() + "' is neither a plugin or a rule source and cannot be applied.");
             } else {
-                boolean imperative = plugin.isImperative();
                 Runnable adder = addPluginInternal(plugin);
                 if (adder != null) {
-                    if (imperative) {
-                        Plugin<?> pluginInstance = producePluginInstance(pluginClass);
-                        instances.put(pluginClass, pluginInstance);
-
-                        if (plugin.isHasRules()) {
-                            applicator.applyImperativeRulesHybrid(pluginIdStr, pluginInstance);
-                        } else {
-                            applicator.applyImperative(pluginIdStr, pluginInstance);
-                        }
-
-                        // Important not to add until after it has been applied as there can be
-                        // plugins.withType() callbacks waiting to build on what the plugin did
-                        pluginContainer.add(pluginInstance);
-                    } else {
-                        applicator.applyRules(pluginIdStr, pluginClass);
-                    }
-
-                    adder.run();
+                    buildOperationExecutor.run("Apply plugin " + plugin.getDisplayName(), new AddPluginAction(plugin, adder));
```

Comment:
I understand your suggestion as getting the display name from the action if it implements `Describable`. The api actually require a display name always, I'm not sure it would be for the better api-wise.

We always consume the display name internally in `DefaultBuildOperationExecutor` to build the internal operation's details object.

Except for the internal details object, the only case where we won't consume the display name is when the operation is not progress-logged and there's no tapi progress listener and no build scan plugin is applied.